From: IrfanMohammad <irfan@graphiant.com>
Date: Thu, 27 Jul 2023 12:46:22 +0000
Subject: add tests for module change subscription recovery

Add a test for a process dying with a connection held that has a
module change subscription to test the module change subscription
recovery logic and particularly interactions with producer xpath
filtering.
---
 tests/test_process.c | 107 +++++++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 107 insertions(+)

diff --git a/tests/test_process.c b/tests/test_process.c
index b547bd7..4c0f8c6 100644
--- a/tests/test_process.c
+++ b/tests/test_process.c
@@ -26,6 +26,8 @@
 #include <unistd.h>
 
 #include "sysrepo.h"
+#include "common.h"
+
 #include "tests/tcommon.h"
 
 #define sr_assert(cond) if (!(cond)) { TLOG_ERR("\"%s\" not true", #cond); sr_assert_line(); abort(); }
@@ -1036,6 +1038,110 @@ test_sub(int rp, int wp)
     return 0;
 }
 
+/* TEST */
+struct state_cb_s {
+    sr_conn_ctx_t *conn;
+    ATOMIC_T cb_called;
+};
+
+static int
+test_change_done_cb(sr_session_ctx_t *session, uint32_t sub_id, const char *module_name, const char *xpath,
+        sr_event_t event, uint32_t request_id, void *private_data)
+{
+    struct state_cb_s *st = (struct state_cb_s *)private_data;
+
+    (void)session;
+    (void)sub_id;
+    (void)module_name;
+    (void)xpath;
+    (void)event;
+    (void)request_id;
+
+    ATOMIC_INC_RELAXED(st->cb_called);
+
+    return SR_ERR_OK;
+}
+
+static int
+test_change_done_crash2(int rp, int wp)
+{
+    sr_subscription_ctx_t *sub = NULL, *sub2 = NULL;
+    sr_session_ctx_t *sess_sub = NULL;
+    sr_session_ctx_t *sess;
+    int ret = 0, r;
+    struct state_cb_s st = { .cb_called = 0 };
+    int count;
+
+    r = sr_connect(0, &st.conn);
+    sr_assert_true(r == SR_ERR_OK);
+
+    barrier(rp, wp);
+
+    r = sr_session_start(st.conn, SR_DS_OPERATIONAL, &sess_sub);
+    sr_assert_true(r == SR_ERR_OK);
+
+    r = sr_module_change_subscribe(sess_sub, "ietf-interfaces", "/ietf-interfaces:interfaces/interface/description", test_change_done_cb, &st, 500, SR_SUBSCR_DONE_ONLY, &sub);
+    sr_assert_true(r == SR_ERR_OK);
+
+    r = sr_module_change_subscribe(sess_sub, "ietf-interfaces", NULL, test_change_done_cb, &st, 1000, SR_SUBSCR_DONE_ONLY, &sub2);
+    sr_assert_true(r == SR_ERR_OK);
+
+    r = sr_session_start(st.conn, SR_DS_OPERATIONAL, &sess);
+    sr_assert_true(r == SR_ERR_OK);
+
+    r = sr_set_item_str(sess, "/ietf-interfaces:interfaces/interface[name='eth1']/type", "iana-if-type:ethernetCsmacd", NULL, 0);
+    sr_assert_true(r == SR_ERR_OK);
+    r = sr_apply_changes(sess, 0);
+    sr_assert_true(r == SR_ERR_OK);
+
+    count = 0;
+    while ((ATOMIC_LOAD_RELAXED(st.cb_called) < 1) && (count < 1500)) {
+        usleep(10000);
+        ++count;
+    }
+    sr_assert_int_equal(ATOMIC_LOAD_RELAXED(st.cb_called), 1);
+
+    sr_unsubscribe(sub2);
+    sr_unsubscribe(sub);
+    sr_session_stop(sess_sub);
+    sr_session_stop(sess);
+    sr_disconnect(st.conn);
+    return ret;
+}
+
+static int
+test_change_done_crash1(int rp, int wp)
+{
+    int ret;
+    sr_conn_ctx_t *conn;
+    sr_subscription_ctx_t *sub = NULL;
+    sr_session_ctx_t *sess;
+
+    ret = sr_connect(0, &conn);
+    sr_assert_true(ret == SR_ERR_OK);
+
+    ret = sr_session_start(conn, SR_DS_OPERATIONAL, &sess);
+    sr_assert_true(ret == SR_ERR_OK);
+
+    ret = sr_module_change_subscribe(sess, "ietf-interfaces", "/ietf-interfaces:interfaces/interface/description", dummy_change_cb, NULL, 500, SR_SUBSCR_NO_THREAD | SR_SUBSCR_DONE_ONLY, &sub);
+    sr_assert_true(ret == SR_ERR_OK);
+
+    barrier(rp, wp);
+
+    exit(0);
+
+    /* will block until crash */
+    while (1) {
+        sr_subscription_process_events(sub, NULL, NULL);
+    }
+
+    /* unreachable */
+    sr_unsubscribe(sub);
+    sr_session_stop(sess);
+    sr_disconnect(conn);
+    return 0;
+}
+
 int
 main(void)
 {
@@ -1048,6 +1154,7 @@ main(void)
         {"context change", test_context_change, test_context_change_sub, setup, teardown},
         {"conn create", test_conn_create, test_conn_create, setup, teardown},
         {"sub apply", test_sub, test_apply, setup, teardown},
+        {"change done crash", test_change_done_crash1, test_change_done_crash2, setup, teardown},
     };
 
     test_log_init();
